Automattic/jetpack
# Connection Package: Handle disconnections gracefully

At the moment, the `jetpack-connection` package doesn’t properly handle disconnections when multiple plugins use the connection. Any plugin that utilizes `jetpack-connection` package relies on the same set of tokens as Jetpack itself, meaning that disconnect call by any plugin leads to the connection being severed to all other plugins, including Jetpack.

#### Changes proposed in this Pull Request:
The PR introduces following improvements:
- Initialization of the connection via `Jetpack_Config` now allows providing connection options.
- Classes `Plugin` and `Plugin_Storage` that handle the plugin connections.
- Changes in the `Manager` class that check if the disconnecting plugin is the only active connection, and don't let it remove the connection otherwise.
- Jetpack plugin provides its information to the connection manager.
- Jetpack "Disconnect" modal modified to display the list of connected plugins.
- Unit tests for `Plugin`, `Plugin_Storage` and `Manager` classes, including new function mocks to imitate WordPress behavior.
- Adjustment of the existing package code to make it in line with WordPress coding standards.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* `p1HpG7-8VJ-p2`

#### Testing instructions:
1. Create a file `wp-content/plugins/jetpack-connection-test-1.php` with following content:
```
<?php
/** Plugin Name: Jetpack Connection Test Plugin (1) */
add_action( 'plugins_loaded', function() {
	( new Automattic\Jetpack\Config() )->ensure( 'connection', array( 'slug' => 'jetpack-connection-test-1', 'name' => 'Jetpack Connection Test Plugin (1)' ) );
}, 1 );
```
2. Go to the "Plugins" section and enable that plugin.
3. Go to the "Jetpack -> Dashboard" and click on "Connections -> Disconnect".
Make sure the modal window displays that new plugin in the list:
<img width="400" alt="jetpack-disconnect-plugins" src="https://user-images.githubusercontent.com/1341249/80242778-89182c80-862b-11ea-8ffe-12e485d1d90f.png">

4. Go to "Plugins" and disable that plugin.

5. Go back to the "Jetpack -> Dashboard" (or make sure to refresh the tab if you left that open), click on "Connections -> Disconnect".
Make sure the modal window no longer displays the "Jetpack Connection is also used by other plugin" section.

6. Run unit tests over the `packages/connection` directory and make sure no failures or errors are reported: `cd packages/connection && composer install && vendor/bin/phpunit`

#### Proposed changelog entry for your changes:
* "Disconnect Jetpack" modal window displays the list of other plugins that use the connection.
